src/python/gem5/resources/resource.py

```diff
@@ -654,47 +662,56 @@ def obtain_resource(
-            resource_directory = os.getenv(
-                "GEM5_RESOURCE_DIR", _get_default_resource_dir()
-            )
+        # If the `to_path` parameter is set, we use that as the path to which
+        # the resource is to be downloaded. Otherwise, default to the
+        # `resource_directory` parameter plus the resource ID.
+        if not to_path:
+            # If the `resource_directory` parameter is not set via this
+            # function, we heck the "GEM5_RESOURCE_DIR" environment variable.
+            # If this too is not set we call `_get_default_resource_dir()` to
+            # determine where the resource directory is, or should be, located.
+            if resource_directory == None:
+                resource_directory = os.getenv(
+                    "GEM5_RESOURCE_DIR", _get_default_resource_dir()
+                )
 
-        # Small checks here to ensure the resource directory is valid.
-        if os.path.exists(resource_directory):
-            if not os.path.isdir(resource_directory):
-                raise Exception(
-                    "gem5 resource directory, "
-                    "'{}', exists but is not a directory".format(
-                        resource_directory
+            # Small checks here to ensure the resource directory is valid.
+            if os.path.exists(resource_directory):
+                if not os.path.isdir(resource_directory):
+                    raise Exception(
+                        "gem5 resource directory, "
+                        "'{}', exists but is not a directory".format(
+                            resource_directory
+                        )
                     )
-                )
-        else:
-            # `exist_ok=True` here as, occasionally, if multiple instance of
-            # gem5 are started simultaneously, a race condition can exist to
-            # create the resource directory. Without `exit_ok=True`, threads
-            # which lose this race will thrown a `FileExistsError` exception.
-            # `exit_ok=True` ensures no exception is thrown.
-            os.makedirs(resource_directory, exist_ok=True)
 
-        # This is the path to which the resource is to be stored.
-        to_path = os.path.join(resource_directory, resource_id)
+            # This is the path to which the resource is to be stored.
+            to_path = os.path.join(resource_directory, resource_id)
+
+        assert to_path
+
+        # Here we ensure the directory in which the resource is to be stored
+        # is created.
+        #
+        # `exist_ok=True` here as, occasionally, if multiple instance of gem5
+        # are started simultaneously, a race condition can exist to create the
+        # resource directory. Without `exit_ok=True`, threads which lose this
+        # race will thrown a `FileExistsError` exception. `exit_ok=True`
+        # ensures no exception is thrown.
+        Path(to_path).parent.mkdir(parents=True, exist_ok=True)
```

Comment:
You may want to wrap this in a try/except and print a nice error if it fails. E.g., if the user specifies a `to_path` that they don't have permission for it would be nice to not have a backtrace but instead have a `fatal`. 

This is a minor nit. Feel free to ignore.

Reply:
I've added a helpful `fatal` here.